### ISSUE TITLE
InfoGAN loss function seems to be wrong

### ISSUE THREAD

**Issue by petertulala** [tdeboissiere/DeepLearningImplementations#47]
Why do you use categorical cross-entropy for loss function of categorical latent code in InfoGAN? The [paper](https://arxiv.org/pdf/1606.03657.pdf) uses lower bound of mutual information (entropy minus conditional entropy), i.e.:

> I(c, G(z,c)) = H(c) - H(c | G(z,c))

But **cross-entropy** H(c, Q(c | G(z,c))) between input and output latent codes instead of **conditional entropy** seems to me it's not correct.

**Comment by petertulala** [comment]
My fault, using cross-entropy is totally fine, good job. It's shown later in paper that H(c, Q(c | G(z,c))) is in fact the lower bound for H(c | G(z,c)).

**Comment by petertulala** [comment]
Reopening, I reviewed the Gaussian loss function, but I think it's not correct. This is how log-likelihood of Gaussian should be calculated:
![screen shot 2017-09-24 at 02 08 13](https://user-images.githubusercontent.com/613623/30778123-7328abc0-a0cd-11e7-8998-7e4ef07cc25f.png)

And this is how the Gaussian loss is calculated in `train.py`:
![screen shot 2017-09-24 at 02 13 26](https://user-images.githubusercontent.com/613623/30778138-f733fed8-a0cd-11e7-8983-982c619bfb6c.png)

The `log 2pi` may be skipped, since it is a constant, but I believe the rest is wrong. Correct implementation:
```
def gaussian_loss(y_true, y_pred):
    mean = y_true[:, 0, :]
    log_stdev = y_true[:, 1, :]
    x = y_pred[:, 0, :]

    frac = K.square(x - mean) / (K.exp(log_stdev) + K.epsilon())
    return 0.5 * K.mean(np.log(2 * np.pi) + log_stdev + frac)
```

I tested the fixed equation and this is the result:
![download 2](https://user-images.githubusercontent.com/613623/30778174-049548d8-a0cf-11e7-8565-93952e1efa2a.png)

See also the implementation from [OpenAI](https://github.com/openai/InfoGAN/blob/master/infogan/misc/distributions.py) which calculates the loss the same way.

**Comment by tdeboissiere** [comment]
log(s**2) = 2 log(s).
Hence the 0.5 factor goes away (1/2 * log(s**2) = log(s)).
So I believe the current implementation is correct. It does ignore the constant factor as you noted.

Your new implementation is correct as well if you treat log_stdev as log(s**2).

**Comment by petertulala** [comment]
Thank you for answer. However in in that case, shouldn't it be multiplied by `0.25` instead of `0.5`?

<img width="576" alt="screen shot 2017-09-26 at 10 07 30" src="https://user-images.githubusercontent.com/613623/30849409-8c709efe-a2a2-11e7-80a4-f85186f5711e.png">

I think if we ignore the constant, the correct loss should be:

    epsilon = (y_true - Q_C_mean) / (K.exp(Q_C_logstd) + K.epsilon())
    loss_Q_C = (Q_C_logstd + 0.25 * K.square(epsilon))

**Comment by tdeboissiere** [comment]
We have:

log((2pi s^2)^(-0.5) * e^(-0.5(x-mu)^2 / s^2)) =
cst + log(s^(-1)) - 0.5 (x - mu) ^ 2 / s^2 =
cst - (log(s) + 0.5 ((x - mu) / e^(log(s))) ^ 2

Which is the expression in my code

**Comment by petertulala** [comment]
Ok thanks, now I see both ways are correct, the difference is only in skipped constant. I made a mistake in my previous comment, it should have been just `σ^2` instead of `log σ^2` in denominator.
